odoo/tools/populate.py

```diff
@@ -59,12 +61,26 @@ def get_field_variation_date(model: Model, field: Field, factor: int, series_ali
     setting duplicated records too far back in the past.
     """
     total_days = min((MAX_DATETIME - MIN_DATETIME).days, factor)
-    return SQL("""
-        %(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days'
-    """, field=SQL.identifier(field.name),
+    cast_type = SQL('date') if field.type == 'date' else SQL('timestamp')
+    query = SQL("""
+        (%(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days')::%(cast_type)s
+    """, field=SQL('value::%s', SQL(cast_type)) if field.company_dependent else SQL.identifier(field.name),
          factor=factor,
          series_alias=SQL.identifier(series_alias),
-         total_days=total_days)
+         total_days=total_days,
+         cast_type=cast_type)
+    if field.company_dependent:
```

Comment:
This function is way more complex than necessary.  You have to combine _both_ `if field.company_dependent` in order to understand each piece separately.  Please make it simple and stupid:
```py
def get_field_variation_date(...):
    total_days = ...
    cast_type = ...
    if not field.company_dependent:
        # base case
        return SQL(...)
    # company-dependent case
    return SQL(...)
```

Reply:
You prefer repeating the block
```sql
sql_expr = SQL("""
        (%(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days')::%(cast_type)s
    """, field=SQL('value::%s', cast_type) if field.company_dependent else SQL.identifier(field.name),
         factor=factor,
         series_alias=SQL.identifier(series_alias),
         total_days=total_days,
         cast_type=cast_type)
```
for each branching? once with field := field.name, the other time value + conversion?
I initially wanted to avoid the repetition, but if you prefer I will change it.

Reply:
I think I've found a good compromise between readability while avoiding duplication for the base SQL expression. Let me know what you think about it.